### ISSUE TITLE
Literals list in `syntax/case` and `syntax/parse` have odd effects on source locations.

### ISSUE THREAD

**Issue by LeifAndersen** [racket/racket#1733]
tl;dr Why aren't literal identifiers in a pattern bound as pattern variables?

- - -

Lets say I have some function `foo` which accepts a syntax list beginning with `add1`, and returns the source location of that `add1`. (Making sure that `add1` is the literal `add1` as defined in `racket/base`.)

Two possible ways to implement this function seem to have different results:

1. Use the literals list (either with `syntax-case` or with `#:literals` in `syntax/parse`
2. Don't use the literals list, but check if the two identifiers are `free-identifier=?` in the guard clause.

The resulting file should look something like this:

```
#lang racket

(require syntax/srcloc)

(define (foo stx)
  (syntax-case stx (add1)
    [(add1 x)
     (build-source-location #'add1)]))

(define add1-identifier #'add1)
(define (foo2 stx)
  (syntax-case stx ()
    [(add1 x)
     (free-identifier=? #'add1 add1-identifier)
     (build-source-location #'add1)]))

(foo #'(add1 5))
(foo2 #'(add1 5))
```

Where `foo`uses the literals list, and `foo2` uses `free-identifer=?`. I would expect the resulting source locations to be one line apart from each other. However, when I run this code, I get:

```
(srcloc 'unsaved-editor 8 30 128 4)
(srcloc 'unsaved-editor 18 9 339 4)
```

Which is to say, the first `add1` is coming from inside the function, while the second case is the actual location. It is worth noting I get the exact same behavior when I use `syntax/parse` instead of `syntax/case`.

What I suspect is happening here is that the `add1` in the first function is testing for identifier equality with the `add1` given in the literals list. Once it matches though, it does not bind `add1` as a pattern variable as I would expect.  In the second case, since there is no identifiers list, `add1` becomes a pattern variable, and thus using it in the body works as expected.

This hypothesis can also be validated by clicking `Check Syntax` in DrRacket. The different `add1`s are colored differently.

While this explanation makes sense to me, it does not seem like the correct behavior. I would expect something in the literals list to check the relevant identifiers for equality, if they don't match, reject the pattern, otherwise continue as normal (aka bind the identifier to a pattern variable).

**Comment by samth** [comment]
I think your understanding of the semantics is correct, but this has been the behavior for a very long time and seems hard to change compatibly.

**Comment by LeifAndersen** [comment]
Are you aware of any programs that rely on the literals list _not_ binding identifiers they match as pattern variables? (I'm not, but I've also not been around as long as you have.)

**Comment by samth** [comment]
This program wouldn't work under the obvious way of implementing your suggestion:

```
(syntax-case #'(+ + +) (+)
   [(+ + +) 1] 
   [_ 2])
```

I've certainly written patterns that use elements of the literals list more than once.

**Comment by LeifAndersen** [comment]
AH, okay. Fair point. I guess its not clear which pattern to bind if you use a literal multiple times. Never mind then. I'll close this issue.

**Comment by rfindler** [comment]
You could maybe add a variant on #:literals that did bind them?

Robby

**Comment by LeifAndersen** [comment]
Possibly. Although it might make make more sense to do this with syntax classes though. Honestly, I'm fine either way.

**Comment by rmculpepper** [comment]
`syntax-parse` lets you use literals like syntax classes, so you can write the following:

    (define-syntax (foo stx)
      (syntax-parse stx #:literals (add1)
        [(a:add1 x)
         (build-source-location #'a)]))

----

Here's another example of a problem caused by binding literals as pattern variables:

    (syntax-case stx (+)
      [(+ e ...)
       (andmap constant? (syntax->list #'(e ...)))
       #`(quote #,(apply + (constants->values #'(e ...))))])

That code would be illegal, because the pattern variable binding would shadow the `+` procedure.
